# Data Wrangling

In this notebook, we will perform essential data wrangling tasks to prepare the dataset for further analysis. These steps are critical in ensuring data quality, integrity, and consistency before moving to Exploratory Data Analysis (EDA).

### Objectives:

- Identify and remove duplicate records from the dataset  
- Detect missing values in the dataset  
- Impute missing values using appropriate techniques  
- Normalize relevant features to bring them onto a similar scale  

These tasks help ensure that our insights are driven by accurate, complete, and clean data — which is a foundational step in any real-world analytics project.


### Import Required Library

We begin by importing the `pandas` library, which is essential for data manipulation and wrangling tasks throughout this notebook.


In [33]:
import pandas as pd

We load the dataset from a CSV file into a pandas DataFrame for inspection and cleaning.


In [34]:
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/VYPrOu0Vs3I0hKLLjiPGrA/survey-data-with-duplicate.csv")

# Display the First 5 row
df.head()

ResponseId                      MainBranch                 Age  \
0           1  I am a developer by profession  Under 18 years old   
1           2  I am a developer by profession     35-44 years old   
2           3  I am a developer by profession     45-54 years old   
3           4           I am learning to code     18-24 years old   
4           5  I am a developer by profession     18-24 years old   

            Employment RemoteWork   Check  \
0  Employed, full-time     Remote  Apples   
1  Employed, full-time     Remote  Apples   
2  Employed, full-time     Remote  Apples   
3   Student, full-time        NaN  Apples   
4   Student, full-time        NaN  Apples   

                                    CodingActivities  \
0                                              Hobby   
1  Hobby;Contribute to open-source projects;Other...   
2  Hobby;Contribute to open-source projects;Other...   
3                                                NaN   
4                                                NaN   

                                             EdLevel  \
0                          Primary/elementary school   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3  Some college/university study without earning ...   
4  Secondary school (e.g. American high school, G...   

                                           LearnCode  \
0                             Books / Physical media   
1  Books / Physical media;Colleague;On the job tr...   
2  Books / Physical media;Colleague;On the job tr...   
3  Other online resources (e.g., videos, blogs, f...   
4  Other online resources (e.g., videos, blogs, f...   

                                     LearnCodeOnline  ... JobSatPoints_6  \
0                                                NaN  ...            NaN   
1  Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
2  Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
3  Stack Overflow;How-to videos;Interactive tutorial  ...            NaN   
4  Technical documentation;Blogs;Written Tutorial...  ...            NaN   

  JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
0            NaN            NaN            NaN             NaN   
1            0.0            0.0            0.0             0.0   
2            NaN            NaN            NaN             NaN   
3            NaN            NaN            NaN             NaN   
4            NaN            NaN            NaN             NaN   

  JobSatPoints_11           SurveyLength SurveyEase ConvertedCompYearly JobSat  
0             NaN                    NaN        NaN                 NaN    NaN  
1             0.0                    NaN        NaN                 NaN    NaN  
2             NaN  Appropriate in length       Easy                 NaN    NaN  
3             NaN               Too long       Easy                 NaN    NaN  
4             NaN              Too short       Easy                 NaN    NaN  

[5 rows x 114 columns]

### Identify Duplicate Rows

We check for duplicate entries in the dataset, count how many exist, and display the first few to understand their structure before removal.


In [35]:
# Check for Duplicates
num_duplicates = df.duplicated().sum()
print(f"Number of duplicate rows : {num_duplicates}")

# display the first few duplicate records
df[df.duplicated()].head()

Number of duplicate rows : 20


ResponseId                      MainBranch                 Age  \
65437           1  I am a developer by profession  Under 18 years old   
65438           2  I am a developer by profession     35-44 years old   
65439           3  I am a developer by profession     45-54 years old   
65440           4           I am learning to code     18-24 years old   
65441           5  I am a developer by profession     18-24 years old   

                Employment RemoteWork   Check  \
65437  Employed, full-time     Remote  Apples   
65438  Employed, full-time     Remote  Apples   
65439  Employed, full-time     Remote  Apples   
65440   Student, full-time        NaN  Apples   
65441   Student, full-time        NaN  Apples   

                                        CodingActivities  \
65437                                              Hobby   
65438  Hobby;Contribute to open-source projects;Other...   
65439  Hobby;Contribute to open-source projects;Other...   
65440                                                NaN   
65441                                                NaN   

                                                 EdLevel  \
65437                          Primary/elementary school   
65438       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
65439    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
65440  Some college/university study without earning ...   
65441  Secondary school (e.g. American high school, G...   

                                               LearnCode  \
65437                             Books / Physical media   
65438  Books / Physical media;Colleague;On the job tr...   
65439  Books / Physical media;Colleague;On the job tr...   
65440  Other online resources (e.g., videos, blogs, f...   
65441  Other online resources (e.g., videos, blogs, f...   

                                         LearnCodeOnline  ... JobSatPoints_6  \
65437                                                NaN  ...            NaN   
65438  Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
65439  Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
65440  Stack Overflow;How-to videos;Interactive tutorial  ...            NaN   
65441  Technical documentation;Blogs;Written Tutorial...  ...            NaN   

      JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
65437            NaN            NaN            NaN             NaN   
65438            0.0            0.0            0.0             0.0   
65439            NaN            NaN            NaN             NaN   
65440            NaN            NaN            NaN             NaN   
65441            NaN            NaN            NaN             NaN   

      JobSatPoints_11           SurveyLength SurveyEase ConvertedCompYearly  \
65437             NaN                    NaN        NaN                 NaN   
65438             0.0                    NaN        NaN                 NaN   
65439             NaN  Appropriate in length       Easy                 NaN   
65440             NaN               Too long       Easy                 NaN   
65441             NaN              Too short       Easy                 NaN   

      JobSat  
65437    NaN  
65438    NaN  
65439    NaN  
65440    NaN  
65441    NaN  

[5 rows x 114 columns]

### Analyze Characteristics of Duplicates

Before removing duplicates, we investigate whether certain columns — such as `MainBranch`, `Employment`, and `RemoteWork` — frequently contain identical values within duplicate rows.

This helps us understand the nature of redundancy in the data and decide if we should drop full-row duplicates or consider subset-based duplication.


In [36]:
# check for duplicates based on selected subset of columns
subset_duplicates = df.duplicated(subset=['MainBranch', 'Employment', 'RemoteWork'])
num_subset_duplicates = subset_duplicates.sum()
print(f"Number of duplicates based on selected columns: {num_subset_duplicates}")

# display a few of those rows
df[subset_duplicates].head()


Number of duplicates based on selected columns: 64896


ResponseId                      MainBranch                 Age  \
1            2  I am a developer by profession     35-44 years old   
2            3  I am a developer by profession     45-54 years old   
14          15  I am a developer by profession     25-34 years old   
15          16  I am a developer by profession     45-54 years old   
16          17     I code primarily as a hobby  Under 18 years old   

             Employment                            RemoteWork   Check  \
1   Employed, full-time                                Remote  Apples   
2   Employed, full-time                                Remote  Apples   
14  Employed, full-time  Hybrid (some remote, some in-person)  Apples   
15  Employed, full-time  Hybrid (some remote, some in-person)  Apples   
16   Student, full-time                                   NaN  Apples   

                                     CodingActivities  \
1   Hobby;Contribute to open-source projects;Other...   
2   Hobby;Contribute to open-source projects;Other...   
14  Hobby;Professional development or self-paced l...   
15                                              Hobby   
16                                                NaN   

                                              EdLevel  \
1        Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2     Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
14    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
15  Some college/university study without earning ...   
16                          Primary/elementary school   

                                            LearnCode  \
1   Books / Physical media;Colleague;On the job tr...   
2   Books / Physical media;Colleague;On the job tr...   
14  Books / Physical media;Other online resources ...   
15  Books / Physical media;On the job training;Oth...   
16  Books / Physical media;Other online resources ...   

                                      LearnCodeOnline  ... JobSatPoints_6  \
1   Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
2   Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
14  Stack Overflow;Coding sessions (live or record...  ...            NaN   
15  Technical documentation;Books;Written Tutorial...  ...            0.0   
16  Technical documentation;Written Tutorials;Stac...  ...            NaN   

   JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
1             0.0            0.0            0.0             0.0   
2             NaN            NaN            NaN             NaN   
14            NaN            NaN            NaN             NaN   
15            0.0            0.0            0.0             0.0   
16            NaN            NaN            NaN             NaN   

   JobSatPoints_11           SurveyLength                  SurveyEase  \
1              0.0                    NaN                         NaN   
2              NaN  Appropriate in length                        Easy   
14             NaN  Appropriate in length                        Easy   
15             0.0  Appropriate in length  Neither easy nor difficult   
16             NaN  Appropriate in length                        Easy   

   ConvertedCompYearly JobSat  
1                  NaN    NaN  
2                  NaN    NaN  
14                 NaN    NaN  
15                 NaN    5.0  
16                 NaN    NaN  

[5 rows x 114 columns]

### Interpretation of Subset-Based Duplicates

We found 64,896 duplicate rows based on `MainBranch`, `Employment`, and `RemoteWork`. This pattern is expected in survey data, where many respondents share similar roles and work settings.

Since other columns like `Age`, `EdLevel`, `ConvertedCompYearly`, and `JobSat` may contain unique insights, we’ll keep these rows for now and only remove exact full-row duplicates.

This ensures we retain valuable variation for deeper analysis.


## Strategic Removal of Duplicates

We define uniqueness in this dataset primarily using the `ResponseId` column. If not unique, a combination of `MainBranch`, `Employment`, `RemoteWork`, and `ConvertedCompYearly` is used to identify and remove duplicate records without losing meaningful data.


In [37]:
df["ResponseId"].is_unique


False

In [38]:
df["ResponseId"].duplicated().sum()


20

Since `ResponseId` and a basic composite key didn’t ensure uniqueness, we expanded our key to include subjective fields like `SurveyLength` and `SurveyEase`. This helped better isolate true duplicates for accurate cleanup.


In [39]:
df.duplicated(subset=['MainBranch', 'Employment', 'RemoteWork', 'Age', 'ConvertedCompYearly', 'SurveyLength', 'SurveyEase']).sum()


42328



Despite testing multiple combinations of profile fields, a significant number of near-identical rows were found. To maintain data integrity, we dropped full-row duplicates from the dataset, keeping only the first occurrence of each unique response.


## Removing duplicates

To ensure each record in the dataset is unique and to avoid inflated statistics, we removed all fully duplicated rows from the DataFrame.


In [40]:
df.drop_duplicates(inplace = True)

After dropping duplicate rows, we verified the cleanup by counting the number of duplicates remaining in the dataset to ensure data integrity.


In [41]:
df.duplicated().sum()

0

## Remove Unnecessary Columns

Dropped non-analytical columns such as `ResponseId`, `Check`, and granular job satisfaction rating columns (`JobSatPoints_1` to `JobSatPoints_11`) to streamline the dataset for analysis.


In [42]:
# Drop ID and survey control columns
df.drop(columns=['ResponseId', 'Check'], inplace=True)

# Drop detailed job satisfaction point columns
for i in range(1, 12):
    col_name = f'JobSatPoints_{i}'
    if col_name in df.columns:
        df.drop(columns=[col_name], inplace=True)


## Identifying Missing Values

We begin by detecting missing values across all columns.  
This helps assess data quality and decide how to handle incomplete entries in the next steps.


In [43]:
# Display total missing values for each column
missing_values = df.isnull()
missing_values

MainBranch    Age  Employment  RemoteWork  CodingActivities  EdLevel  \
0           False  False       False       False             False    False   
1           False  False       False       False             False    False   
2           False  False       False       False             False    False   
3           False  False       False        True              True    False   
4           False  False       False        True              True    False   
...           ...    ...         ...         ...               ...      ...   
65432       False  False       False       False             False    False   
65433       False  False       False       False             False     True   
65434       False  False       False       False             False    False   
65435       False  False       False       False             False    False   
65436       False  False       False        True              True     True   

       LearnCode  LearnCodeOnline  TechDoc  YearsCode  ...  TimeAnswering  \
0          False             True     True       True  ...           True   
1          False            False    False      False  ...           True   
2          False            False    False      False  ...           True   
3          False            False     True      False  ...           True   
4          False            False    False      False  ...           True   
...          ...              ...      ...        ...  ...            ...   
65432      False             True     True      False  ...           True   
65433       True             True     True       True  ...           True   
65434      False            False    False      False  ...           True   
65435      False            False    False      False  ...           True   
65436       True             True     True       True  ...           True   

       Frustration  ProfessionalTech  ProfessionalCloud  ProfessionalQuestion  \
0             True              True               True                  True   
1             True              True               True                  True   
2             True              True               True                  True   
3             True              True               True                  True   
4             True              True               True                  True   
...            ...               ...                ...                   ...   
65432         True              True               True                  True   
65433         True              True               True                  True   
65434         True              True               True                  True   
65435         True              True               True                  True   
65436         True              True               True                  True   

       Industry  SurveyLength  SurveyEase  ConvertedCompYearly  JobSat  
0          True          True        True                 True    True  
1          True          True        True                 True    True  
2          True         False       False                 True    True  
3          True         False       False                 True    True  
4          True         False       False                 True    True  
...         ...           ...         ...                  ...     ...  
65432      True          True        True                 True    True  
65433      True          True        True                 True    True  
65434      True          True        True                 True    True  
65435      True          True        True                 True    True  
65436      True          True        True                 True    True  

[65437 rows x 103 columns]

In [44]:
# Counting missing values for all columns
for column in missing_values.columns.values.tolist():
    print(column)
    print(missing_values[column].value_counts())
    print(" ")

MainBranch
MainBranch
False    65437
Name: count, dtype: int64
 
Age
Age
False    65437
Name: count, dtype: int64
 
Employment
Employment
False    65437
Name: count, dtype: int64
 
RemoteWork
RemoteWork
False    54806
True     10631
Name: count, dtype: int64
 
CodingActivities
CodingActivities
False    54466
True     10971
Name: count, dtype: int64
 
EdLevel
EdLevel
False    60784
True      4653
Name: count, dtype: int64
 
LearnCode
LearnCode
False    60488
True      4949
Name: count, dtype: int64
 
LearnCodeOnline
LearnCodeOnline
False    49237
True     16200
Name: count, dtype: int64
 
TechDoc
TechDoc
False    40897
True     24540
Name: count, dtype: int64
 
YearsCode
YearsCode
False    59869
True      5568
Name: count, dtype: int64
 
YearsCodePro
YearsCodePro
False    51610
True     13827
Name: count, dtype: int64
 
DevType
DevType
False    59445
True      5992
Name: count, dtype: int64
 
OrgSize
OrgSize
False    47480
True     17957
Name: count, dtype: int64
 
PurchaseInfluence
Pur

We will select a column with significant missing data (`EdLevel`,`Employment`, `RemoteWork`) and fill missing entries with the most frequent value to maintain consistency.


In [45]:
print("EdLevel")
print(df["EdLevel"].value_counts().head())

EdLevel
EdLevel
Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                          24942
Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                       15557
Some college/university study without earning a degree                                 7651
Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)     5793
Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                         2970
Name: count, dtype: int64


In [46]:
# Find the most frequent value in the RemoteWork column
most_frequent_edlevel = df['EdLevel'].mode()[0]
print(f"Most frequent value in EdLevel column: {most_frequent_edlevel}")

# Fill missing values in EdLevel with the most frequent value
most_frequent_edlevel = df['EdLevel'].mode()[0]
df['EdLevel'].fillna(most_frequent_edlevel)



Most frequent value in EdLevel column: Bachelor’s degree (B.A., B.S., B.Eng., etc.)


0                                Primary/elementary school
1             Bachelor’s degree (B.A., B.S., B.Eng., etc.)
2          Master’s degree (M.A., M.S., M.Eng., MBA, etc.)
3        Some college/university study without earning ...
4        Secondary school (e.g. American high school, G...
                               ...                        
65432         Bachelor’s degree (B.A., B.S., B.Eng., etc.)
65433         Bachelor’s degree (B.A., B.S., B.Eng., etc.)
65434         Bachelor’s degree (B.A., B.S., B.Eng., etc.)
65435    Secondary school (e.g. American high school, G...
65436         Bachelor’s degree (B.A., B.S., B.Eng., etc.)
Name: EdLevel, Length: 65437, dtype: object

In [47]:
print("Employment")
print(df["Employment"].value_counts().head())

Employment
Employment
Employed, full-time                                                         39041
Independent contractor, freelancer, or self-employed                         4846
Student, full-time                                                           4709
Employed, full-time;Independent contractor, freelancer, or self-employed     3557
Not employed, but looking for work                                           2341
Name: count, dtype: int64


In [48]:
# Find the most frequent value in the Employment column
most_frequent_employment = df['Employment'].mode()[0]
print(f"Most frequent value in Employment column: {most_frequent_employment}")

# Fill missing values in Employment with the most frequent value
most_frequent_remote = df['Employment'].mode()[0]
df['Employment'].fillna(most_frequent_employment)


Most frequent value in Employment column: Employed, full-time


0        Employed, full-time
1        Employed, full-time
2        Employed, full-time
3         Student, full-time
4         Student, full-time
                ...         
65432    Employed, full-time
65433    Employed, full-time
65434    Employed, full-time
65435    Employed, full-time
65436     Student, full-time
Name: Employment, Length: 65437, dtype: object

In [49]:
print('RemoteWork')
print(df["RemoteWork"].value_counts())

RemoteWork
RemoteWork
Hybrid (some remote, some in-person)    23015
Remote                                  20831
In-person                               10960
Name: count, dtype: int64


In [50]:
# Find the most frequent value in the RemoteWork column
most_frequent_remote = df['RemoteWork'].mode()[0]
print(f"Most frequent value in RemoteWork column: {most_frequent_remote}")

# Fill missing values in RemoteWork with the most frequent value
most_frequent_remote = df['RemoteWork'].mode()[0]
df['RemoteWork'].fillna(most_frequent_remote)



Most frequent value in RemoteWork column: Hybrid (some remote, some in-person)


0                                      Remote
1                                      Remote
2                                      Remote
3        Hybrid (some remote, some in-person)
4        Hybrid (some remote, some in-person)
                         ...                 
65432                                  Remote
65433                                  Remote
65434                               In-person
65435    Hybrid (some remote, some in-person)
65436    Hybrid (some remote, some in-person)
Name: RemoteWork, Length: 65437, dtype: object

### Imputation Completed

The missing values in the `EdLevel` , `Employment` and `RemoteWok` column were successfully filled using the most frequent value.  
Post-imputation, no missing values remain in this column.


## Normalizing Compensation Data



in this step, we will Apply Min-Max Scaling and store results in a new column `ConvertedCompYearly_MinMax`.


In [65]:
# Step 1: Filter out rows where ConvertedCompYearly is not null
df_comp = df[df['ConvertedCompYearly'].notna()].copy()

# Step 2: Reset index (optional, for clean analysis)
df_comp.reset_index(drop=True, inplace=True)

# Step 3: Min-Max Normalization
min_val = df_comp['ConvertedCompYearly'].min()
max_val = df_comp['ConvertedCompYearly'].max()
df_comp['ConvertedCompYearly_MinMax'] = (df_comp['ConvertedCompYearly'] - min_val) / (max_val - min_val)
df_comp['ConvertedCompYearly_MinMax'].value_counts().head()



ConvertedCompYearly_MinMax
0.003998    42039
0.003964      321
0.003303      308
0.004625      230
0.005285      226
Name: count, dtype: int64

  
we will now apply Z-score method to Standardize the `ConvertedCompYearly` column and save results in a new column `ConvertedCompYearly_Zscore`.


In [63]:
# Z-score normalization for ConvertedCompYearly
mean_val = df_comp['ConvertedCompYearly'].mean()
std_val = df_comp['ConvertedCompYearly'].std()
df_comp['ConvertedCompYearly_Zscore'] = (df_comp['ConvertedCompYearly'] - mean_val) / std_val
df_comp['ConvertedCompYearly_Zscore'].value_counts().head()


ConvertedCompYearly_Zscore
-0.067513    42039
-0.072468      321
-0.168181      308
 0.023237      230
 0.118950      226
Name: count, dtype: int64

### Save the Dataset into CSV file

In [66]:
df.to_csv("Cleaned_dataset.csv", index = False)

---